torch/csrc/jit/interpreter.h

```diff
@@ -44,18 +43,52 @@ struct TORCH_API Code {
   friend std::ostream & operator<<(std::ostream & out, const Code & code);
 };
 
+// InterpreterState state that and used to compute a Code
```

Comment:
Ha, there must be a reason for that: `InterpreterState::pImpl` is of type `c10::intrusive_ptr<InterpreterStateImpl>`. `c10::intrusive_ptr` requires the full definition of its target (`InterpreterStateImpl`). So that's why I put most functions of `InterpreterStateImpl` private and leave its constructor public.

Reply:
Ok, it looks like intrusive_ptr isn't implemented correctly because it should not require the definition of InterpreterStateImpl to exist for its functions to be defined. 

In that case, we should file a bug against c10::intrusive_ptr, and in the mean time, use `std::shared_ptr< InterpreterStateImpl>` and  `std::enabled_shared_from_this` instead.

Reply:
(it would be much worse to expose the implementation in the header file as it is,  it defeats the purpose of the pImpl pattern being used).

Reply:
in person: we can use `c10::intrusive_ptr<c10::intrusive_ptr_target>` in the header file and then static down cast in the implementation.

Reply:
Also, as far as I know, this is an intrinsic issue with any intrusive pointers and the pImpl pattern because there is no way to communicate that the pImpl type must inherit from c10::intrusive_ptr_target without defining it completely, which is exactly what the pImpl pattern is trying to avoid.